src/app/ReadHandler.cpp

```diff
@@ -52,25 +51,13 @@ void ReadHandler::Shutdown()
 {
     InteractionModelEngine::GetInstance()->ReleaseClusterInfoList(mpAttributeClusterInfoList);
     InteractionModelEngine::GetInstance()->ReleaseClusterInfoList(mpEventClusterInfoList);
-    AbortExistingExchangeContext();
+    mpExchangeCtx = nullptr;
     MoveToState(HandlerState::Uninitialized);
-    mpDelegate                 = nullptr;
     mpAttributeClusterInfoList = nullptr;
     mpEventClusterInfoList     = nullptr;
     mCurrentPriority           = PriorityLevel::Invalid;
 }
 
-CHIP_ERROR ReadHandler::AbortExistingExchangeContext()
-{
-    if (mpExchangeCtx != nullptr)
-    {
-        mpExchangeCtx->Abort();
-        mpExchangeCtx = nullptr;
-    }
-
-    return CHIP_NO_ERROR;
-}
-
 CHIP_ERROR ReadHandler::OnReadRequest(Messaging::ExchangeContext * apExchangeContext, System::PacketBufferHandle && aPayload)
```

Comment:
This method is unconditionally overwriting `mpExchangeCtx`.  Isn't this a possible way to leak an exchange context?

Reply:
Yes, if someone ever called `OnReadRequest` on a `ReadHandler` that has a non-null `mpExchangeCtx` that it's keeping open.  In practice that should not happen because whenever we are keeping the `mpExchangeCtx` open we are in a state other than `Uninitialized` and hence will not get an `OnReadRequest`.

We could add a check that we're in the `Uninitialized` state at the beginning of `OnReadRequest`... In practice this API is meant to be used from one place only (`InteractionModelEngine::OnReadRequest`) and that use is guarded on exactly this state check.

The _right_ way to do this is to use constructors and destructors in the read handler allocation codepath and make the exchange context a constructor arg.  Or at least moving the setting of the exchange context to `Init`, where it's maybe clearer that it happens exactly once in the "life" of a `ReadHandler`.  Again I would like to avoid scope-creeping this PR and do that sort of thing in a followup if that's OK with you.

Reply:
I think you're referring to the 'IsFree' check in this code:

```
    for (auto & readHandler : mReadHandlers)
    {
        if (readHandler.IsFree())
        {
            err = readHandler.Init(mpDelegate);
            SuccessOrExit(err);
            err               = readHandler.OnReadRequest(apExchangeContext, std::move(aPayload));
            apExchangeContext = nullptr;
            break;
        }
    }
```

Yes, I'm a little weary of this type of code that so manually manages lifecycles for things like the `ReadHandler` object.  I realize that we need to provide statically allocated pools for memory-constrained targets.  But that does not necessarily mean we must have these semantics where we manually find a thing in an array that 'is free', and then execute public methods on it that are only safe if it is free.

It seems like it would be far safer to use some sort of malloc / free interfaces where you can only get a handle to a `ReadHandler` in the first place if it is free.

This would also allow us to more easily move to dynamically allocated pools.  As it stands, are we really constrained to `CHIP_MAX_NUM_READ_HANDLER` (4) in the entire stack?  That is going to be a really big problem on some platforms.

OK, I will chalk this up to something that needs a larger refactor.  So no need to block this PR necessarily.

Reply:
>  As it stands, are we really constrained to `CHIP_MAX_NUM_READ_HANDLER (4)` in the entire stack?

Yes, we are, and this is clearly not ok for all devices....

I filed https://github.com/project-chip/connectedhomeip/issues/8650 to improve this stuff.

Reply:
Thank you!